src/RegistrationManager.ts

```diff
@@ -1,549 +1,569 @@
+    if (newValue !== oldValue) {
+      log.info(`Manager stake is ${newValue ? 'now' : 'no longer'} locked`);
+      this.printNotRegisteredMessage();
+    }
+  }
+
+  constructor(
+    contractInteractor: ContractInteractor,
+    transactionManager: TransactionManager,
+    txStoreManager: TxStoreManager,
+    eventEmitter: EventEmitter,
+    config: ServerConfigParams,
+    // exposed from key manager?
+    managerAddress: string,
+    workerAddress: string
+  ) {
+    const listener = (): void => {
+      this.printNotRegisteredMessage();
+    };
+    this._balanceRequired = new AmountRequired(
+      'Balance',
+      BigNumber.from(config.blockchain.managerMinBalance),
+      listener
+    );
+    this._stakeRequired = new AmountRequired(
+      'Stake',
+      BigNumber.from(config.blockchain.managerMinStake),
+      listener
+    );
+
+    this._contractInteractor = contractInteractor;
+    this._hubAddress = config.contracts.relayHubAddress;
+    this._managerAddress = managerAddress;
+    this._workerAddress = workerAddress;
+    this._eventEmitter = eventEmitter;
+    this._transactionManager = transactionManager;
+    this._txStoreManager = txStoreManager;
+    this._config = config;
+  }
+
+  async init(): Promise<void> {
+    if (this._lastWorkerAddedTransaction == null) {
+      this._lastWorkerAddedTransaction =
+        await this._queryLatestWorkerAddedEvent();
     }
 
-    /**
-     * @param withdrawManager - whether to send the relay manager's balance to the owner.
-     *        Note that more than one relay process could be using the same manager account.
-     * @param currentBlock
-     */
-    async withdrawAllFunds(
-        withdrawManager: boolean,
-        currentBlock: number
-    ): Promise<PrefixedHexString[]> {
-        let transactionHashes: PrefixedHexString[] = [];
-        transactionHashes = transactionHashes.concat(
-            await this._sendWorkersEthBalancesToOwner(currentBlock)
-        );
-        if (withdrawManager) {
-            transactionHashes = transactionHashes.concat(
-                await this._sendManagerEthBalanceToOwner(currentBlock)
-            );
-        }
-
-        this.eventEmitter.emit('unstaked');
-        return transactionHashes;
+    this._isInitialized = true;
+  }
+
+  async handlePastEvents(
+    hubEventsSinceLastScan: TypedEvent[],
+    lastScannedBlock: number,
+    currentBlock: number,
+    forceRegistration: boolean
+  ): Promise<string[]> {
+    if (!this._isInitialized) {
+      throw new Error('RegistrationManager not initialized');
+    }
+    const options = {
+      fromBlock: lastScannedBlock + 1,
+      toBlock: 'latest',
+    };
+
+    type DefaultManagerEvent = Extract<
+      ManagerEvent,
+      'StakeAdded' | 'StakeUnlocked' | 'StakeWithdrawn'
+    >;
+    const eventsNames: DefaultManagerEvent[] = [
+      'StakeAdded',
+      'StakeUnlocked',
+      'StakeWithdrawn',
+    ];
+    // FIXME the relayManager needs to be send as parameter
```

Comment:
Its a blocker, it works but it does not work correctly if we have more than 1 server. Since we are removing the ContractInteractor it would require some major changes.

Reply:
updated